### PR TITLE
use rocksdb sst file ingestion for snapshot

### PR DESCRIPTION
**Pull request by hhkbp2**
Hi,

This PR tries to fix issue #927.

### CODE REVIEW

**Comment by hhkbp2** [comment]
PTAL @siddontang @BusyJay

**Comment by ngaut** [comment]
Please fix CI.

**Comment by siddontang** [comment]
Seem we don't support old snapshot format now. 
If we upgrade TiKV, but a region's state is applying and the snapshot file is the old format, how can we handle it?

File: src/raftstore/store/worker/metrics.rs [hunk 1]
```diff
@@ -75,12 +75,6 @@ lazy_static! {
             "Bucketed histogram of raftstore hash compution duration"
         ).unwrap();
 
-    pub static ref SNAPSHOT_KV_COUNT_HISTOGRAM: Histogram =
-        register_histogram!(
-            "tikv_snapshot_kv_count",
```
Comment (siddontang) [inline src/raftstore/store/worker/metrics.rs]: why remove this?

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -870,6 +867,54 @@ fn build_snap_file(f: &mut SnapFile,
     Ok(())
 }
 
+pub fn encode_cf_files_size(sizes: Vec<(String, u64)>) -> Vec<KeyValue> {
+    let mut kvs = vec![];
+    for (cf_name, size) in sizes {
+        let cf_size = format!("{}", size);
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: why not using varint or 8 bytes codec directly?

File: src/raftstore/store/peer_storage.rs [hunk 3]
```diff
@@ -870,6 +867,54 @@ fn build_snap_file(f: &mut SnapFile,
...
+    }
+    kvs
+}
+
+pub fn decode_cf_files_size(kvs: &[KeyValue]) -> Result<Vec<(String, u64)>> {
+    let mut cf_sizes: Vec<(String, u64)> = vec![];
+    for kv in kvs {
+        let key = kv.get_key();
+        let cf = match str::from_utf8(key) {
+            Ok(s) => s.to_string(),
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: to_owned?

File: src/raftstore/store/snap.rs [hunk 4]
```diff
@@ -1,25 +1,48 @@
...
+const SST_FILE_SUFFIX: &'static str = ".sst";
+
+fn get_snapshot_cfs() -> Vec<String> {
+    let size = ALL_CFS.len();
+    let mut cfs = Vec::with_capacity(size);
+    for cf in ALL_CFS {
+        if *cf == CF_RAFT {
+            continue;
+        }
+        cfs.push(String::from(*cf));
```
Comment (siddontang) [inline src/raftstore/store/snap.rs]: cf.to_owned()

File: src/raftstore/store/snap.rs [hunk 5]
```diff
@@ -66,218 +89,543 @@ const SNAP_GEN_PREFIX: &'static str = "gen";
...
+        true
     }
 
-    /// Get a validation reader.
-    pub fn reader(&self) -> io::Result<SnapValidationReader> {
-        SnapValidationReader::open(self.path())
+    pub fn add_kv(&mut self, k: &[u8], v: &[u8]) -> io::Result<()> {
+        let mut cf_file = &mut self.cf_files[self.index];
+        let mut writer = cf_file.writer.as_mut().unwrap();
+        if let Err(e) = writer.add(k, v) {
```
Comment (siddontang) [inline src/raftstore/store/snap.rs]: can we use `try!` here directly?

**Comment by hhkbp2** [comment]
The old format should be supported rather than removed.

**Comment by ngaut** [comment]
Coooooool

**Comment by siddontang** [comment]
This PR is too huge now, can we split it?

**Comment by ngaut** [comment]
We can split it into two PRs. @hhkbp2 
The last commit can be a single PR.

**Comment by siddontang** [comment]
how do we do if the new TiKV sends a new snapshot to the old TiKV?

Seem that the old TiKV will save the file with the old format, but it is not correct.

**Comment by hhkbp2** [comment]
@siddontang 
The old code doesn't check the version of snapshot format. The upgrade process could be handled d like a two-phase procedure:
1. the binary of new code could be upgraded first with new format disabled
2. after all binaries are upgrade, the new format could be enabled

**Comment by siddontang** [comment]
So I think we should use a config to control using SST snapshot file or not.

**Comment by ngaut** [comment]
Is is possible that we stop balance when upgrade to new tikv ? after we finish the upgrade process, we enable balancing.

**Comment by siddontang** [comment]
No, PD can't control the snapshot sending if  there is a log gap between the leader and follower.

**Comment by hhkbp2** [comment]
@ngaut @siddontang 
I'm going to separate this PR into two PRs
1. Abstract the interface for adding a new format
2. add the new format impl
What do you think?

**Comment by siddontang** [comment]
+1

**Comment by ngaut** [comment]
Good idea.

**Comment by hhkbp2** [comment]
This PR is split into #1579 #1618. Close.

**Pull request closed without merging**